src/test/scala/is/hail/expr/ir/TableIRSuite.scala

```diff
@@ -47,4 +49,193 @@ class TableIRSuite extends SparkSuite {
+    (17,    1,  -1,    1,    1),
+    (17,    1,  -1,    2,    1),
+    (17,    2,  -1,    1,    1),
+    (17,    2,  -1,    2,    1),
+    (18, null,null,    1,    1),
+    (18, null,null,    2,    1),
+    (21, null,null,    1,    1),
+    (21, null,null,    2,    1),
+    (22,    1,  -1,    1,    1),
+    (22,    1,  -1,    2,    1),
+    (22,    2,  -1,    1,    1),
+    (22,    2,  -1,    2,    1),
+    (23,    1,  -1, null, null),
+    (23,    2,  -1, null, null),
+    (26,    1,  -1, null, null),
+    (26,    2,  -1, null, null),
+    (27,    1,  -1,    1,    1),
+    (27,    1,  -1,    2,    1),
+    (27,    2,  -1,    1,    1),
+    (27,    2,  -1,    2,    1),
+    (28, null,null,    1,    1),
+    (28, null,null,    2,    1),
+    (31, null,null,    1,    1),
+    (31, null,null,    2,    1),
+    (32,    1,  -1,    1,    1),
+    (32,    1,  -1,    2,    1),
+    (32,    2,  -1,    1,    1),
+    (32,    2,  -1,    2,    1),
+    (33,    1,  -1, null, null),
+    (33,    2,  -1, null, null),
+    (36,    1,  -1, null, null),
+    (36,    2,  -1, null, null),
+    (37,    1,  -1,    1,    1),
+    (37,    1,  -1,    2,    1),
+    (37,    2,  -1,    1,    1),
+    (37,    2,  -1,    2,    1),
+    (38, null,null,    1,    1),
+    (38, null,null,    2,    1),
+    (41, null,null,    1,    1),
+    (41, null,null,    2,    1)
+  ).map(Row.fromTuple)
+
+  val leftPartitioners = Array(
+    IndexedSeq(
+      Interval(Row(0, 0), Row(4, 1), true, false),
+      Interval(Row(10, -1), Row(19, 1), true, false),
+      Interval(Row(20, 0), Row(24, 0), true, true),
+      Interval(Row(25, 0), Row(39, 0), true, true))
+//    IndexedSeq(
+//      Interval(Row(0, 0), Row(10), true, false),
+//      Interval(Row(10), Row(44, 0), true, true)),
+//    IndexedSeq(Interval(Row(), Row(), true, true))
+  ).map(new OrderedRVDPartitioner(kType, _))
+
+  val rightPartitioners = Array(
+    IndexedSeq(
+      Interval(Row(5, 0), Row(9, 1), true, false),
+      Interval(Row(15, -1), Row(29, 1), true, false),
+      Interval(Row(30, 0), Row(34, 0), true, true),
+      Interval(Row(35, 0), Row(44, 0), true, true))
+//    IndexedSeq(
+//      Interval(Row(0, 0), Row(10), true, false),
+//      Interval(Row(10), Row(44, 0), true, true)),
+//    IndexedSeq(Interval(Row(), Row(), true, true))
+  ).map(new OrderedRVDPartitioner(kType, _))
+
+  val joinTypes = Array(
+    ("outer", (row: Row) => true),
+    ("left",  (row: Row) => !row.isNullAt(1)),
+    ("right", (row: Row) => !row.isNullAt(3)),
+    ("inner", (row: Row) => !row.isNullAt(1) && !row.isNullAt(3))
+  )
+
+  @DataProvider(name = "join")
+  def joinData(): Array[Array[Any]] =
+    for {
+      l <- leftPartitioners
+      r <- rightPartitioners
+      (j, p) <- joinTypes
+      leftProject <- Seq[Set[Int]](Set(), Set(1), Set(2), Set(1, 2))
+      rightProject <- Seq[Set[Int]](Set(), Set(1), Set(2), Set(1, 2))
+      if !leftProject.contains(1) || rightProject.contains(1)
```

Comment:
Right, the exceptional case is `leftProject.contains(1) && !rightProject.contains(1)`, i.e. we projected out the extra left key field (so the left key has duplicates now), but we didn't project out the extra right key field.

By the way, I have a plan to make joins work in all cases, including this one, but I think I should leave that for a separate PR, because it needs changes to the iterator join level.